### PR TITLE
OCI: Enhance oci implementation to include oci node pools

### PR DESCRIPTION
**Pull request by trungng92**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This pull request enhances the cluster autoscaler's current oci (Oracle Cloud Infrastructure) cloud provider implementation to work with oci node pools in addition to working with oci instance pools.

#### Which issue(s) this PR fixes:
None. This is a new feature. OCI currently provides a close source implementation for node pools, but this pull requests open sources this implementation.

#### Special notes for your reviewer:
In OCI, there are two concepts, instance pools and node pools. Instance pools are generic instances. Node pools are specifically a set of instances in a pool used as Kubernetes nodes.

In this implementation, only one type of pool is accepted. This means that only instance pools **xor** node pools will be accepted. If the user provides multiple types of pools, this will exit out with a relevant error statement. In the future, we will support mixed instance pools and node pools.

#### Does this PR introduce a user-facing change?
Users will now be able to use the open source version of the cluster autoscaler with Oracle Cloud Infrastructure's Container Engine for Kubernetes Node Pools.

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/oci/OWNERS
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/OWNERS:3]: I think we should be more restrictive with the number of approvers. Also, there is a _requirement_ that approvers (and reviewers) be [Kubernetes org members](https://github.com/kubernetes/community/blob/master/community-membership.md). There is also an expectation that approvers be involved with the autoscaling project (e.g. monitoring for provider issues, regularly attending Kubernetes SIG-Autoscaling, etc.).

File: cluster-autoscaler/cloudprovider/oci/oke/oci/cache.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/oke/oci/cache.go:5]: I see the `oci` package name is duplicated. For maintainability, we should organize the implementation specific code in separate packages (`oke` & `instancepools`). 

There is also common code between the two implementations that could live an a common package (`common`) so that we could re-use certain functions, constants, etc. rather than having duplicate code (`oci_util.go` is a good start).

File: cluster-autoscaler/cloudprovider/oci/oke/oci/oci_ref.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/oke/oci/oci_ref.go:23]: This function and even this annotation string are both examples of code-reuse opportunities. 

`const ociAnnotationCompartmentID string = "oci.oraclecloud.com/compartment-id"` is is defined in `oci_cloud_provider.go` and the now two separate implementations of `nodeToOciRef` should be combined into one shared function.

File: cluster-autoscaler/cloudprovider/oci/oke/oci/oci_cloud_provider.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/oke/oci/oci_cloud_provider.go:135]: Configuration parameters that make sense for both implementations should be honored. Our [README](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/oci/README.md#optional-cloud-config-file) documents the parameters that can be set via `--cloud-config`. However, the OKE implementation is dropping the config here: `CreateOciManager(nil, do, createKubeClient(opts))`.

File: cluster-autoscaler/cloudprovider/oci/oke/oci/oci_manager.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/oke/oci/oci_manager.go:335]: However, we _do_ wait in the instancepool implementation - at least for the instances to be provisioned. 

We also check for capacity and quota errors so that the autoscaler to try to scale a different pool (rather than waiting for `max-node-provision-time` and then trying to scale the same pool again).

We may decide that certain functional behaviors are acceptable between the two implementations. However, they should probably be considered on a case by case basis.

**Review by jlamillan — CHANGES_REQUESTED** [review]
Looks like there is still some work to do with this PR to ensure the behaviors & configurations of the two implementations are consistent. Also, we probably have some more initial refactoring to do in terms of code-organization and reuse to do.

File: cluster-autoscaler/cloudprovider/oci/oke/oci/cache.go
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/oke/oci/cache.go:5]: Yes, I agree that there is a lot of common code between the two implementations. There is a plan afterwards to integrate the two implementations more tightly. This is an initial pass to open source our internal code that we know is working.

File: cluster-autoscaler/cloudprovider/oci/oke/oci/oci_manager.go
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/oke/oci/oci_manager.go:337]: > However, we do wait in the instancepool implementation - at least for the instances to be provisioned.

I'll set up a meeting with relevant people to discuss how we want to handle this.

> We also check for capacity and quota errors so that the autoscaler to try to scale a different pool (rather than waiting for max-node-provision-time and then trying to scale the same pool again). We added this stuff based on customer feedback.

I do not know if Node Pools provide an easy way to know whether it is failing based on these criteria. But these are a good ideas to add as a future enhancement.

File: cluster-autoscaler/cloudprovider/oci/OWNERS
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/oci/OWNERS:3]: We (SIG-Autoscaling) don't apply any necessary contribution count threshold to become a reviewer/approver of company-specific code. As long as there is some indication that a person is employed at the company, we are fine with adding them, trusting that the company internal org structure will handle the rest. So we are fine with both long and short list here. Anyway, as @jlamillan is the current approver, he has the final say in this topic.

**Comment by trungng92** [comment]
> Looks like there is still some work to do with this PR to ensure the behaviors & configurations of the two implementations are consistent. Also, we probably have some more initial refactoring to do in terms of code-organization and reuse to do.

I've made changes relevant to the previous comments. Mainly, I've created 3 different packages:

1. package to contain instance pool code
2. package to contain node pool code
3. shared common package

File: cluster-autoscaler/cloudprovider/oci/oke/oci/oci_manager.go
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/oke/oci/oci_manager.go]: I had a discussion with our team regarding adding a wait to the node pool implementation. We have decided not to add a wait to the node pool implementation as the current implementation does not do this, and we did not want to introduce any kind of potential behavior change.

In particular, adding a wait for a scale up that is quickly followed by a scale down should not cause any issues because the cluster autoscaler already has logic that puts the scale down into a backoff for X minutes after a scale up is performed.

However, in the case where a scale up is fast followed by another scale up, introducing a wait may mean that the autoscaler will have to wait for the previous scale to complete before it can make an additional decision to scale up again.

File: cluster-autoscaler/cloudprovider/oci/oke/oci/oci_cloud_provider.go
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/oke/oci/oci_cloud_provider.go]: completed with current refactor

File: cluster-autoscaler/cloudprovider/oci/oke/oci/oci_ref.go
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/oke/oci/oci_ref.go]: completed with current refactor

File: cluster-autoscaler/cloudprovider/oci/oke/oci/cache.go
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/oke/oci/cache.go]: completed with current refactor. See latest comment. I have split things into three compartments.

1. instance pool
2. node pool
3. common

File: cluster-autoscaler/cloudprovider/builder/builder_all.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/builder/builder_all.go:133]: It _might_ be better to import `instancepools` aliased as "oci" so that we can keep the keep the return `cloudprovider.Build` format. 

i.e. 

```
import (
	oci "k8s.io/autoscaler/cluster-autoscaler/cloudprovider/oci/instancepools"
)
```
...
```
	case cloudprovider.OracleCloudProviderName:
		return oci.BuildOCI(opts, do, rl)
```

File: cluster-autoscaler/cloudprovider/oci/README.md
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/README.md:141]: Probably want to note that `OCI_USE_NON_POOL_MEMBER_ANNOTATION` applies to instance pools only.

File: cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go:90]: I'm using `InstancePrincipalConfigurationProviderForRegion` in the instance pool implementation and therefore _require_ `OCI_REGION` be set when `OCI_USE_INSTANCE_PRINCIPAL` is set to `true` (see [note in README](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/oci/README.md#environment-variables)).

If both implementations can use `InstancePrincipalConfigurationProvider` we should update the instance pool implementation so the implementations are consistent wrt configuring Instance Principals (and remove the note in the `README`).

File: cluster-autoscaler/cloudprovider/builder/builder_oci.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/builder/builder_oci.go:39]: It _might_ be better to import `instancepools` aliased as "oci" so that we can keep the keep the return `cloudprovider.Build` format. 

i.e. 

```
import (
	oci "k8s.io/autoscaler/cluster-autoscaler/cloudprovider/oci/instancepools"
)
```
...
```
	case cloudprovider.OracleCloudProviderName:
		return oci.BuildOCI(opts, do, rl)
```

File: cluster-autoscaler/cloudprovider/oci/README.md
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/README.md:145]: Is `XXX` intended to mean that the node pool does not support any environment variables? Or was this a TODO that didn't get updated?

As far as I can tell, _both_ implementations support `OCI_USE_INSTANCE_PRINCIPAL` and `OCI_REFRESH_INTERVAL` environment variables. I added another comment asking whether if the call to `InstancePrincipalConfigurationProviderForRegion` in `oci_instance_pool_manager.go` could be replaced with `InstancePrincipalConfigurationProvider`. If so, the two implementations would behave similarly and the need for the `OCI_REGION` environment variable could go away. That would potentially leave only `OCI_USE_NON_POOL_MEMBER_ANNOTATION` as the only environment variable that only had use in one implementation. The others would be common to both.

File: cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go:580]: This function is _almost_ identical to `BuildGenericLabelsForInstancePool` in `oci_instance_pool_manager.go` except that the instance pool implementation appears to be missing support for checking for and adding the Ampere / ARM labels if the shape matches. 

Could you add the simple ARM shape check to the instance pool implementation and possibly move the `ArmArch` constant and regular expression into the common package?

File: cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go:393]: If you could add the following line to the instance-pool implementation (`oci_instance_pool_manager.go` `GetInstancePoolNodes`) the behaviors would be consistent: 

```
@@ -261,6 +264,8 @@ func (m *InstancePoolManagerImpl) GetInstancePoolNodes(ip InstancePoolNodeGroup)
        for _, instance := range *instanceSummaries {
                status := &cloudprovider.InstanceStatus{}
                switch *instance.State {
+               case string(core.InstanceLifecycleStateStopped), string(core.InstanceLifecycleStateTerminated):
+                       klog.V(4).Infof("skipping instance is in stopped/terminated state: %q", *instance.Id)
                case string(core.InstanceLifecycleStateRunning):
                        status.State = cloudprovider.InstanceRunning
                case string(core.InstanceLifecycleStateCreatingImage):
```

File: cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go:90]: I see. I was wondering why the OCI_REGION env var was required for instance principals and it's because of the difference in how the configuration provider is initialized. I'll update the instance principal implementation to also use this provider so that OCI_REGION is not needed.

Comment (jlamillan) [reply cluster-autoscaler/cloudprovider/oci/nodepools/oci_manager.go:90]: Updated the `InstancePrincipalConfigurationProvider` in the Instance Pool implementation to be non region specific

File: cluster-autoscaler/cloudprovider/builder/builder_oci.go
Comment (jlamillan) [reply cluster-autoscaler/cloudprovider/builder/builder_oci.go:39]: Added 'oci' alias to Oracle Cloud import to preserve provider naming pattern.

File: cluster-autoscaler/cloudprovider/oci/OWNERS
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/OWNERS]: Discussed with Jesse Millan, I've added me and sohan-oracle to the list of reviewers/approvers (commented out)

File: cluster-autoscaler/cloudprovider/oci/README.md
Comment (trungng92) [reply cluster-autoscaler/cloudprovider/oci/README.md]: We've since updated this to have

```
#### Environment-variables applicable _only_ for Node Pools
n/a
```

**Review by jlamillan — APPROVED** [review]
Thanks Trung. You and I have discussed some followup work offline, which we can continue to iterate on (including looking into existing whether the implementations should leverage the External gRPC provider mechanism).

For now, this pull-request LGTM.
 

/lgtm

**Comment by towca** [comment]
The OCI part was LGTMd by an OCI owner, the rest LGTM.

/approve

**Pull request merged**
